Add static HtmlUtils/StringUtils into edx namespace

andy-armstrong opened the pull request:
This change loads key UI Toolkit classes into the ``edx`` namespace so that they can be consumed by parts of the platform that cannot use RequireJS yet (mostly XBlocks and XModules). This was straightforward in Studio because the classes can be loaded by RequireJS and be guaranteed that they load before any other code. In the LMS that isn't possible, so there the code relies on the following UI Toolkit PR that lets classes load into the ``edx`` namespace if RequireJS is not available: https://github.com/edx/edx-ui-toolkit/pull/46

### Sandbox
- [ ] https://andy-armstrong.sandbox.edx.org/dashboard

### Testing
 - [ ] manual testing on devstack and my sandbox

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens
- [ ] Code review: @efischer19   

FYI: @robrap @AlasdairSwan @bjacobel @tobz @peter-fogg 

### Post-review
- [x] Squash commits

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

andy-armstrong commented on common/static/js/src/string_utils.js:
```diff
@@ -45,4 +47,277 @@
     };
     this.interpolate_text = interpolate_text;
+
+    edx.StringUtils = (function() {
```
Reviewers: I put this into string_utils as it already exists and is available in both Studio and LMS. Ultimately we need to find a way to load the UI Toolkit classes asynchronously and then have them made available in the edx namespace in the same way, but that seems difficult because LMS and Studio behave differently, and because we need to ensure that the client code is never invoked before the files have completed loading.

@tobz may have a good insight as to how to do this.

robrap replied:
Add the Temporary clone comment here again.

robrap commented:
:+1:  with the additional comment.  Thanks.

peter-fogg commented on common/static/js/src/string_utils.js:
```diff
@@ -45,4 +47,277 @@
+         * this would look like:
+         *
+         *   HtmlUtils.interpolate(
+         *       gettext('You are enrolling in {spanStart}{courseName}{spanEnd}'),
```
Should this be "in Django" or "in Javascript"?

peter-fogg commented:
Looks like you'll need some [`jshint: globals`](http://jshint.com/docs/options/#globals) in a a few places; otherwise :+1:

andy-armstrong commented on common/static/common/js/common_libraries.js:
```diff
@@ -1 +1,19 @@
-define(['domReady!', 'jquery', 'backbone', 'underscore', 'gettext']);
+define([
```
Reviewers: it turns out that string_utils is not available in Studio, but @cahrens pointed out that this common_libraries file is where other global namespace variables are set up (primarily ``$`` and ``_``). The nice thing about this solution is that it isn't using a clone of the UI Toolkit code, but is just bringing it in directly as any code would.

andy-armstrong replied:
Given this, I could move the cloned code into an LMS-only directory if folks think that is cleaner.

andy-armstrong replied:
I now use @cahrens' suggested approach for installing the utilities in LMS.

andy-armstrong commented:
Thanks @robrap and @peter-fogg. I've addressed your feedback, cleaned up the quality issues, fixed how the constants are set up in Studio, and added the missing ``StringUtils`` context parameter for ``HtmlUtils.template``. Please re-review when you have a chance.

BTW, I'm not going to merge this until I have successfully written some code that uses the constants and verified on a sandbox that they work in both Studio and LMS.

andy-armstrong commented:
@robrap @peter-fogg I've built a sandbox and verified that both Studio and LMS have the ``edx`` namespace available as expected:

https://andy-armstrong.sandbox.edx.org/

robrap commented:
@andy-armstrong I didn't have a chance to review and have a ton of stuff to look into.  I don't want to be a bottleneck for this.  I'm happy if you and others can use this.  Thanks!

cahrens commented on common/static/js/src/string_utils.js:
```diff
@@ -45,4 +47,278 @@
     };
     this.interpolate_text = interpolate_text;
+
+    edx.StringUtils = edx.StringUtils || (function() {
```
Why is it necessary to copy this code? Given that we only need a solution for LMS (since CMS is covered above), how about putting edx-ui-toolkit/js/utils/string-utils and edx-ui-toolkit/js/utils/html-utils on the window object in lms/static/lms/js/require-config.js (by requiring them after the config section and then putting them on the window object)?

andy-armstrong replied:
@cahrens The primary reason for copying the code was to do something quickly for today's marathon, as it was blocking a number of potential updates. This seemed like the quickest approach to get both Studio and LMS to work (but then I had the realization of the cleaner approach that would work for Studio).

My concern with doing what you suggest is that it is asynchronous, and so any non-RequireJS files could load first and then error trying to access the namespace. In Studio, everything is loaded via RequireJS so we know that nothing will have been invoked before the common libraries have initialized. Having said that, now that the marathon is over there is less urgency so it would make sense to try to come up with a solid approach that doesn't require copying.

cahrens replied:
Good point. I think the right solution is to change HtmlUtils and StirngUtils (in the pattern library) so they install on the window object if RequireJS is not present (using the same syntax that many libraries due, and that the optimizer can handle). Then we list them in base_vendor_js (which means they load BEFORE RequireJS, synchronously) and install them into RequireJS the same way we do with underscore, JQuery, etc.

andy-armstrong commented:
I've been thinking about this comment from @tobz in HipChat:

> it kind of strikes me that we just really need to go all in on RequireJS, or something else, and forego the whole complicated django pipeline nonsense
> so much effort spent on adding two simple helper classes :[

The best option would be to use the same logic to load the helper classes (in common/static/common/js/common_libraries.js) for both Studio and LMS. This already is how things work with Studio, but in LMS the common libraries are added to the page manually through bundles (lms-main-vendor.js et al). I'm wondering what it would take to switch the LMS to load everything in the same way. We didn't do it before because we couldn't get RequireJS Optimizer to understand everything being done on a page, but maybe that isn't necessary now that we can add files to the manual override path. Would it be possible to load RequireJS first, and then load the common_libraries bundle, and only then do all the other JavaScript on the page.

@tobz Is this one part of what you were doing by getting all the JavaScript to load in the footer? How hard would it be just to rejig the page to load the common_libraries first before everything else? Let's talk on Monday about the options here. I need to land some kind of solution here soon so that other folks can land their safe template PRs (@efischer19 is blocked). Maybe I should go with @cahrens' solution since it is very simple, and then we can investigate a cleaner solution.

tobz commented:
I think going with whatever unblocks this most quickly is the right path.

With that said, I've also been thinking a lot about this, too, and my first cogent thought is: what if we generated two RequireJS overrides?

The first would be a programmatically-generated override based on the configuration we use for the pipeline.  That's to say, we would do the same thing the override code was doing before -- take the filename, chop off the `.js`, and then map that as the module name to the file.

We'd then have the manually-specified list of module/file mapping as another override afterwards to handle vendor-specific stuff -- weird extensions, etc.

Now, there's two immediate problems, or regressions in performance, here:

- we'd be generating this dynamically
- we don't want RequireJS loading every single little thing file-by-file, even if they're heavily cached

For the dynamic generation part, my initial thought is that, somehow, we hook into the collectstatic/staticfiles pipeline and sort of cache the result of `<path to dynamically generated RequireJS config>`.  So, it would basically be generating the first time it was requested on a server, then served from disk afterwards, but heavily cached because it was using the hashed naming scheme.  I'm not sure if this is possible without writing some helper code, but who knows.

The second bit is bundled files.  I wonder if there's a way for us to have RequireJS handle the idempotency aspect?  That's to say... your config might have bits that look like:

    'js/vendor/foo/bar': 'static/js/lms-application.7s36dg36dsf',
    'js/vendor/quux/bax': 'static/js/lms-application.7s36dg36dsf'

I wonder, if we did that, if RequireJS would be smart enough to not load the file a second (or third, etc) time if you required both of those modules?  That'd be an interesting way to keep bundled files for performance purposes, but still be able to map every module to the bundle they belong to.  It also, arguably, provides a clearer display when debugging to understand where a module lives. "Oh, the RequireJS config says `js/vendor/quux/baz` is loaded from `lms-application`, got it!"

Thoughts?

cahrens commented:
I typed a response earlier, but it appears to have gotten lost.

Given that LMS is not entirely in RequireJS, I think that trying to get it to behave exactly like Studio is going to be futile. And I don't think we should invest our time trying to convert everything in LMS to RequireJS because I'm not convinced that is the right direction. I think time would be better spent experimenting with React or other server-side rendering options (for performance reasons). 

It's a very common paradigm for libraries to be robust to whether or not an AMD system is present, so my proposal isn't unusual. However, it does raise the question of whether or not the whole UI toolkit would have to use this approach in order to be widely useful in edx-platform.

andy-armstrong commented:
@cahrens @tobz Thanks for the thoughtful comments. Based on both of your advice, I decided to implement @cahrens's simple solution. I updated the UI Toolkit to make the utilities be loadable without RequireJS (see https://github.com/edx/edx-ui-toolkit/pull/46), and then I've updated edx-platform to load these files into the bundles. I verified that both LMS and Studio now get the edx namespace with the expected attributes filled in.

Let me know what you think.

andy-armstrong commented on lms/envs/common.py:
```diff
@@ -1244,6 +1244,13 @@
     'js/vendor/underscore.string.min.js',
+
+    # Make some edX UI Toolkit utilities available in the global "edx" namespace
+    'edx-ui-toolkit/js/utils/global-loader.js',
```
Reviewers: it is slightly odd that these edX files are added to the vendor bundle. However, they have to be loaded before RequireJS, and a few other approaches I took failed. Maybe I should rename the bundles to make it clearer what they are... any thoughts?

cahrens replied:
I'm OK with it, though I realize that we keep adding more bloat to base_vendor (which impacts performance).

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
@robrap @peter-fogg @cahrens  - I've updated my sandbox with the latest implementation suggested by @cahrens. I've verified that both LMS and Studio are bundling JS correctly, and that they both make ``edx.StringUtils`` and ``edx.HtmUtils`` globally available. Please re-review when you have a chance.

Note that I'll update the UI Toolkit version once I publish a new release.

FYI @AlasdairSwan @efischer19

robrap commented:
:+1:  assuming this works as you stated.  I haven't tested this.  I'd rather have @efischer19 ensure that this unblocks his PR.

Also, we still need to update docs accordingly, and determine where we want to do that.  If it makes sense for this all to be under Confluence FedX as well, we can have the Read the Docs page point to it.

Thanks!

cahrens commented:
@andy-armstrong I would have expected to see an update to static/lms/js/require-config.js and lms/js/build.js to define HtmlUtils and StringUtils (given that they will be on the global namespace), ensuring that they don't end up duplicated in other optimized bundles.

As the code currently stands, I don't understand how code is able to reference HtmlUtils and StringUtils via RequireJS in LMS.

andy-armstrong commented:
@cahrens and I spoke, and while my current implementation works, it does end up with the utility classes loaded twice (once into the "edx" namespace and once into page bundles that depend upon the classes). We agreed that it made sense to remove this duplication, so I've implemented her suggestion.

@cahrens @efischer19 Please re-review (and Eric, can you cherry-pick my latest commit to ensure it works for you). Thanks.

andy-armstrong commented:
BTW I've verified the behavior locally on my devstack, and I'm rebuilding my sandbox now.

cahrens commented:
:+1:

andy-armstrong commented:
I've updated to use the hot-off-the-presses 0.9.0 release of the UI Toolkit, and I've squashed my commits. @efischer19 could you cherry-pick one last time to verify that this is good before I go ahead and merge. Thanks!

andy-armstrong commented:
@catong Once this merges, @robrap's safe template documentation will need to be updated to describe when and how to use the new static utilities (available at ``edx.HtmlUtils`` and ``edx.StringUtils``). Let me know how you'd like to do this.

catong commented:
@andy-armstrong I have created DOC-2832 to track the work. I'm happy to help in whatever way I can with editing and getting the content merged. Would you want to start a PR in the doc repo?

andy-armstrong commented:
@catong That sounds good. I'll start a PR.
